x-pack/plugin/repository-encrypted/src/test/java/org/elasticsearch/repositories/encrypted/ChainingInputStreamTests.java

```diff
@@ -0,0 +1,573 @@
+            Assert.assertThat(result[i], Matchers.is(test.v2()[i]));
+        }
+        // middle single empty stream
+        test = testEmptyComponentsInChain(3, Arrays.asList(1));
+        result = test.v1().readAllBytes();
+        assertThat(result.length, Matchers.is(test.v2().length));
+        for (int i = 0; i < result.length; i++) {
+            Assert.assertThat(result[i], Matchers.is(test.v2()[i]));
+        }
+        // leading and trailing empty streams
+        test = testEmptyComponentsInChain(3, Arrays.asList(0, 2));
+        result = test.v1().readAllBytes();
+        assertThat(result.length, Matchers.is(test.v2().length));
+        for (int i = 0; i < result.length; i++) {
+            Assert.assertThat(result[i], Matchers.is(test.v2()[i]));
+        }
+    }
+
+    public void testNullComponentTerminatesChain() throws Exception {
+        TestInputStream[] sourceComponents = new TestInputStream[3];
+        TestInputStream[] chainComponents = new TestInputStream[5];
+        byte[] b1 = new byte[1 + Randomness.get().nextInt(2)];
+        Randomness.get().nextBytes(b1);
+        sourceComponents[0] = new TestInputStream(b1);
+        sourceComponents[1] = null;
+        byte[] b2 = new byte[1 + Randomness.get().nextInt(2)];
+        Randomness.get().nextBytes(b2);
+        sourceComponents[2] = new TestInputStream(b2);
+        ChainingInputStream test = new ChainingInputStream() {
+            int i = 0;
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                chainComponents[i] = (TestInputStream) currentComponentIn;
+                if (i < sourceComponents.length) {
+                    return sourceComponents[i++];
+                } else {
+                    i++;
+                    return null;
+                }
+            }
+
+            @Override
+            public boolean markSupported() {
+                return false;
+            }
+        };
+        byte[] b = test.readAllBytes();
+        assertThat(b.length, Matchers.is(b1.length));
+        for (int i = 0; i < b.length; i++) {
+            Assert.assertThat(b[i], Matchers.is(b1[i]));
+        }
+        assertThat(chainComponents[0], Matchers.nullValue());
+        assertThat(chainComponents[1], Matchers.is(sourceComponents[0]));
+        assertThat(chainComponents[1].closed.get(), Matchers.is(true));
+        assertThat(chainComponents[2], Matchers.nullValue());
+        assertThat(chainComponents[3], Matchers.nullValue());
+    }
+
+    public void testCallsForwardToCurrentComponent() throws Exception {
+        InputStream mockCurrentIn = mock(InputStream.class);
+        ChainingInputStream test = new ChainingInputStream() {
+            @Override
+            InputStream nextComponent(InputStream currentComponentIn) throws IOException {
+                throw new IllegalStateException();
+            }
+        };
+        test.currentIn = mockCurrentIn;
```

Comment:
Per earlier comments, I don't think it necessary to mess with the internal state here - we can just act via the explicit contract.

Reply:
You're right.